### PR TITLE
nautilus: mgr/dashboard: add popover list of Stand-by Managers & Metadata Servers (MDS) in landing page

### PR DESCRIPTION
**Pull request by avanthakkar**
backport tracker: https://tracker.ceph.com/issues/44696

---

backport of https://github.com/ceph/ceph/pull/33529
parent tracker: https://tracker.ceph.com/issues/42979

this backport was staged using ceph-backport.sh version 15.1.0.1009
find the latest version at https://github.com/ceph/ceph/blob/master/src/script/ceph-backport.sh

### CODE REVIEW

**Comment by avanthakkar** [comment]
jenkins test make check

File: src/pybind/mgr/dashboard/controllers/health.py [hunk 1]
```diff
@@ -105,16 +105,25 @@ def df(self):
...
             for fs in fs_map['filesystems']:
                 mdsmap_info = fs['mdsmap']['info']
                 min_mdsmap_info = dict()
                 for k, v in mdsmap_info.items():
+<<<<<<< HEAD
                     min_mdsmap_info[k] = self._partial_dict(v, ['state'])
                 fs['mdsmap'] = dict(info=min_mdsmap_info)
+=======
+                    min_mdsmap_info[k] = partial_dict(v, ['state'])
+>>>>>>> cdfeb1d196... mgr/dashboard: add popover list of managers in landing page
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/controllers/health.py]: Hey @avanthakkar, looks like you haven't properly resolved a git conflict. If you don't know what this mean or how to fix it, please let me know and I'll help.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard/health/health.component.html [hunk 2]
```diff
@@ -97,11 +97,17 @@
     </cd-info-card>
     <cd-info-card cardTitle="Metadata Servers"
                   i18n-cardTitle
+<<<<<<< HEAD
                   class="col-sm-6 col-md-4 col-lg-3"
                   *ngIf="(enabledFeature.cephfs && healthData.fs_map | mdsSummary) as transformedResult"
+=======
+                  class="cd-status-card"
+                  *ngIf="((enabledFeature.cephfs && healthData.fs_map) | mdsSummary) as transformedResult"
+>>>>>>> cdfeb1d196... mgr/dashboard: add popover list of managers in landing page
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard/health/health.component.html]: Same here

File: src/pybind/mgr/dashboard/controllers/health.py (hunk 1 above)
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/controllers/health.py]: You talking about the whitespace added? I think the changes are reverted which I made. You can check with https://github.com/ceph/ceph/pull/33529 , there are few lines which I removed but they are not reflected here. So how can I make the changes and reflect here?

Comment (epuertat) [reply src/pybind/mgr/dashboard/controllers/health.py]: Nope, I mean `<<<<<<< HEAD` or `=======`. These are referred as [merge conflict markers](https://wincent.com/wiki/Understanding_Git_conflict_markers).

Comment (avanthakkar) [reply src/pybind/mgr/dashboard/controllers/health.py]: Yeah I got that. So the changes I made in upstream for this part of code is causing the merge conflict , right?

Comment (avanthakkar) [reply src/pybind/mgr/dashboard/controllers/health.py]: I think the changes I made in upstream branch are not same as that in the backported branch

**Comment by smithfarm** [comment]
@avanthakkar Looking at cdfeb1d196c7d47340baae2be5910b90c889e778 I see that, in the file `src/pybind/mgr/dashboard/controllers/health.py` it makes a change in `mgr_map()`

Can you explain in the commit message why that change was omitted in the cherry-pick?

**Comment by avanthakkar** [comment]
> @avanthakkar Looking at [cdfeb1d](https://github.com/ceph/ceph/commit/cdfeb1d196c7d47340baae2be5910b90c889e778) I see that, in the file `src/pybind/mgr/dashboard/controllers/health.py` it makes a change in `mgr_map()`
> 
> Can you explain in the commit message why that change was omitted in the cherry-pick?

There was some problem while backporting, so the changes didn't get reflect in cherry-pick but were supposed to be made and that's the conflict

**Comment by smithfarm** [comment]
@avanthakkar Would it make sense to manually remove the line from the `mgr_map()` function, so the cherry-pick resembles cdfeb1d more closely?

**Comment by avanthakkar** [comment]
> @avanthakkar Would it make sense to manually remove the line from the `mgr_map()` function, so the cherry-pick resembles [cdfeb1d](https://github.com/ceph/ceph/commit/cdfeb1d196c7d47340baae2be5910b90c889e778) more closely?

Yes, as the fix to be made in mgr_map was to remove that line as it passes the empty list to the API and we don't want the empty list. So while cherry picking due to some issue which I faced, the change didn't get added in that file. So thus to resolve the conflict have to change it manually

**Comment by avanthakkar** [comment]
jenkins test dashboard backend

File: src/pybind/mgr/dashboard/controllers/health.py [hunk 3]
```diff
@@ -106,15 +106,13 @@ def fs_map(self):
...
-            fs_map['standbys'] = [{}] * len(fs_map['standbys'])
             fs_map['filesystems'] = [self._partial_dict(item, ['mdsmap']) for
                                      item in fs_map['filesystems']]
             for fs in fs_map['filesystems']:
                 mdsmap_info = fs['mdsmap']['info']
                 min_mdsmap_info = dict()
                 for k, v in mdsmap_info.items():
                     min_mdsmap_info[k] = self._partial_dict(v, ['state'])
-                fs['mdsmap'] = dict(info=min_mdsmap_info)
         return fs_map
```
Comment (Devp00l) [inline src/pybind/mgr/dashboard/controllers/health.py:117]: Shouldn't line 129 (previous 131) be removed too?

(Just read in the comments that you plan to remove it.)

Comment (avanthakkar) [reply src/pybind/mgr/dashboard/controllers/health.py:117]: Yes , I guess! Thanks for pointing out! Will make the changes

**Comment by avanthakkar** [comment]
jenkins test dashboard backend

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test dashboard backend

**Comment by avanthakkar** [comment]
jenkins test dashboard backend

**Comment by avanthakkar** [comment]
jenkins test dashboard backend

**Comment by callithea** [comment]
- [pkg build](https://shaman.ceph.com/builds/ceph/wip-laura-testing-nautilus-34095-34399-34400/)
- [QA run](http://pulpito.ceph.com/laura-2020-04-08_13:40:25-rados:dashboard-wip-laura-testing-nautilus-34095-34399-34400-distro-basic-smithi/) [WIP]

**Comment by avanthakkar** [comment]
jenkins test dashboard backend

**Comment by callithea** [comment]
jenkins retest this please

**Comment by callithea** [comment]
jenkins test dashboard backend

**Comment by callithea** [comment]
jenkins test dashboard backend

**Comment by callithea** [comment]
[Previous](https://jenkins.ceph.com/job/ceph-dashboard-pr-backend/709/consoleFull) dashboard backend API test failure:
```
2020-04-18 23:17:26,715.715 INFO:__main__:ERROR
2020-04-18 23:17:26,716.716 INFO:__main__:
2020-04-18 23:17:26,716.716 INFO:__main__:======================================================================
2020-04-18 23:17:26,716.716 INFO:__main__:ERROR: setUpClass (tasks.mgr.dashboard.test_monitor.MonitorTest)
2020-04-18 23:17:26,716.716 INFO:__main__:----------------------------------------------------------------------
2020-04-18 23:17:26,717.717 INFO:__main__:Traceback (most recent call last):
2020-04-18 23:17:26,717.717 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/helper.py", line 115, in setUpClass
2020-04-18 23:17:26,717.717 INFO:__main__:    cls._base_uri = cls._get_uri("dashboard").rstrip('/')
2020-04-18 23:17:26,717.717 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/mgr_test_case.py", line 158, in _get_uri
2020-04-18 23:17:26,717.717 INFO:__main__:    cls.wait_until_true(lambda: _get_or_none() is not None, 30)
2020-04-18 23:17:26,717.717 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/ceph_test_case.py", line 164, in wait_until_true
2020-04-18 23:17:26,717.717 INFO:__main__:    raise RuntimeError("Timed out after {0}s".format(elapsed))
2020-04-18 23:17:26,717.717 INFO:__main__:RuntimeError: Timed out after 30s
2020-04-18 23:17:26,717.717 INFO:__main__:
2020-04-18 23:17:26,717.717 INFO:__main__:----------------------------------------------------------------------
2020-04-18 23:17:26,717.717 INFO:__main__:Ran 60 tests in 803.841s
2020-04-18 23:17:26,717.717 INFO:__main__:
2020-04-18 23:17:26,717.717 INFO:__main__:FAILED (errors=1)
2020-04-18 23:17:26,718.718 INFO:__main__:
2020-04-18 23:17:26,718.718 INFO:__main__:======================================================================
2020-04-18 23:17:26,718.718 INFO:__main__:ERROR: setUpClass (tasks.mgr.dashboard.test_monitor.MonitorTest)
2020-04-18 23:17:26,718.718 INFO:__main__:----------------------------------------------------------------------
2020-04-18 23:17:26,718.718 INFO:__main__:Traceback (most recent call last):
2020-04-18 23:17:26,718.718 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/helper.py", line 115, in setUpClass
2020-04-18 23:17:26,718.718 INFO:__main__:    cls._base_uri = cls._get_uri("dashboard").rstrip('/')
2020-04-18 23:17:26,718.718 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/mgr_test_case.py", line 158, in _get_uri
2020-04-18 23:17:26,718.718 INFO:__main__:    cls.wait_until_true(lambda: _get_or_none() is not None, 30)
2020-04-18 23:17:26,718.718 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/ceph_test_case.py", line 164, in wait_until_true
2020-04-18 23:17:26,718.718 INFO:__main__:    raise RuntimeError("Timed out after {0}s".format(elapsed))
2020-04-18 23:17:26,718.718 INFO:__main__:RuntimeError: Timed out after 30s
```

**Comment by callithea** [comment]
@avanthakkar could you please rebase your PR, since there are currently 261 commits in this PR?

**Comment by avanthakkar** [comment]
> @avanthakkar could you please rebase your PR, since there are currently 261 commits in this PR?

Done!

**Comment by avanthakkar** [comment]
jenkins test dashboard backend

**Comment by callithea** [comment]
jenkins test dashboard backend

**Comment by callithea** [comment]
- [QA run](http://pulpito.ceph.com/laura-2020-04-22_14:48:13-rados:dashboard-wip-laura-testing-nautilus-34095-distro-basic-smithi/) [WIP]

**Comment by callithea** [comment]
Latest [QA run](http://pulpito.ceph.com/laura-2020-04-24_11:31:56-rados:dashboard-wip-laura-testing-nautilus-34095-distro-basic-smithi/) [WIP]
There are already failures in some jobs: `Failure Reason:
Test failure: test_minimal_health (tasks.mgr.dashboard.test_health.HealthTest) `

**Comment by avanthakkar** [comment]
jenkins test dashboard backend

**Comment by callithea** [comment]
- [pkg build](https://shaman.ceph.com/builds/ceph/wip-laura-testing-nautilus-34095-2/) [WIP]
- QA run [PENDING]

**Comment by bk201** [comment]
```
(...) != "In `input['fs_map']['standbys'][0]`: unknown keys: set([u'export_targets', u'addr', u'incarnation', u'epoch', u'state_seq', u'state', u'gid', u'flags', u'rank', u'features', u'addrs', u'name'])"
```

Maybe this commit needs to be backported: https://github.com/ceph/ceph/pull/34138/commits/afce413014086ab7cab5de133dae91e30635198a

**Comment by bk201** [comment]
How about just cherry-picking https://github.com/ceph/ceph/commit/afce413014086ab7cab5de133dae91e30635198a into this PR and close https://github.com/ceph/ceph/pull/35494? The changes in https://github.com/ceph/ceph/pull/35494 can't work without changes in this PR.

**Comment by smithfarm** [comment]
> How about just cherry-picking [afce413](https://github.com/ceph/ceph/commit/afce413014086ab7cab5de133dae91e30635198a) into this PR and close #35494? 

@avanthakkar Please feel free to do that, and let me know if you need help to change the PR link on https://tracker.ceph.com/issues/45945

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by callithea** [comment]
- [QA run](http://pulpito.ceph.com/laura-2020-06-22_09:51:14-rados:dashboard-wip-laura-testing-nautilus-34095-35215-35367-distro-basic-smithi/) [WIP]

**Comment by callithea** [comment]
@smithfarm: I think your requested changes wrt the "Conflicts"-section of the commit have been addressed - could you please take a look and dismiss the request if it's fine from your POV?

**Comment by smithfarm** [comment]
@avanthakkar I noticed that the last commit in this PR now has three additional "(cherry picked from commit ...)" lines at the end:

```
(cherry picked from commit 6f1793b)
(cherry picked from commit 83d5c5b)
(cherry picked from commit e0ab29d)
```

Could you amend the commit (`git commit --amend`) and drop these lines, please?

**Comment by avanthakkar** [comment]
> @avanthakkar I noticed that the last commit in this PR now has three additional "(cherry picked from commit ...)" lines at the end:
> 
> ```
> (cherry picked from commit 6f1793b)
> (cherry picked from commit 83d5c5b)
> (cherry picked from commit e0ab29d)
> ```
> 
> Could you amend the commit (`git commit --amend`) and drop these lines, please?

@smithfarm Done!

**Pull request merged**